### PR TITLE
perf!: concurrent merge with optional debugging

### PR DESCRIPTION
**Pull request by mark-rushakoff**
# Background

The non-determinism tests in the SDK have been taking excessively long to run. Profiling the test showed that there are many calls to `codec/types.NewInterfaceRegistry`, which calls `proto.MergedRegistry` in this module. Prior to these optimizations, that method would be limited to a single CPU for about 5ms on my machine when not performing validations, or about 225ms with validations on; with the optimizations, that drops to about 1.5ms or 55ms respectively, on my machine.

While the simulations need to be cleaned up to not call that method so frequently, this was still worth optimizing because:
- other tests will continue to instantiate a new merged registry as needed
- the merged registry is currently required when starting `simd` either in client or server mode. The validations are currently off, but should probably be on as a sanity check -- shaving off 150+ milliseconds of CLI startup time is worth doing for faster automated recovery and for a better CLI experience.

This PR supersedes #56. It differs by offering cleaner control over child goroutines to process file descriptors; backfilling tests and benchmarks; and using an explicitly limited number of goroutines that reuse allocation-heavy values and process data arriving on channels, rather than one goroutine per file descriptor. Thank you to @mmsqe for the work in #56 that influenced this change.

# API change

Calls to `proto.MergedRegistry()` do not need to change.

Existing calls to `proto.MergedFileDescriptors()` need to change to call one of `proto.MergedGlobalFileDescriptors()` or `proto.MergedGlobalFileDescriptorsWithValidation()`. I chose to move the existing `debug` parameter into an explicit part of the method name rather than a boolean argument. This will make call sites clear whether they will fail due to import path mismatches or incorrectly duplicated file descriptors, without the reader needing context on the literal true or false argument.

For finer control over a merge, `proto.MergedFileDescriptors` has been updated to accept arguments `(globalFiles *protoregistry.Files, appFiles map[string][]byte)`. I'm not sure if this will be useful outside of tests, but it was necessary for testing and benchmarking these changes.

# Performance data

In [a draft PR on the SDK](https://github.com/cosmos/cosmos-sdk/pull/15862), this brings the time for `test-sim-nondeterminism`'s test stage down to under 9 minutes: https://github.com/cosmos/cosmos-sdk/actions/runs/4724996838/jobs/8382891823. With the previous implementation, runs taking longer than 45 minutes were common.

Here are benchmarks comparing the previous and new implementations, against a representative but incomplete copy of the protobuf file descriptors as of https://github.com/cosmos/cosmos-sdk/commit/c4489d9785de8b24efe12ba9209669e9dce9e23f:

```
$ go test ./test/mergedregistry/ -cpu=1,2,4,8,10 -bench=Repres -benchtime=5s
goos: darwin
goarch: arm64
pkg: github.com/cosmos/gogoproto/test/mergedregistry
BenchmarkRepresentativeMergedFileDescriptors/sequential            	    1790	   3259606 ns/op	 1636177 B/op	   42145 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/sequential-2          	    2041	   2902895 ns/op	 1636184 B/op	   42145 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/sequential-4          	    2101	   2864636 ns/op	 1636200 B/op	   42145 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/sequential-8          	    2031	   2933869 ns/op	 1636229 B/op	   42145 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/sequential-10         	    2024	   2959004 ns/op	 1636226 B/op	   42145 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/concurrent            	    1783	   3364500 ns/op	 1637314 B/op	   42165 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/concurrent-2          	    3027	   1978660 ns/op	 1694548 B/op	   42176 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/concurrent-4          	    4794	   1301653 ns/op	 1808718 B/op	   42198 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/concurrent-8          	    4257	   1356715 ns/op	 2030442 B/op	   42241 allocs/op
BenchmarkRepresentativeMergedFileDescriptors/concurrent-10         	    4261	   1396325 ns/op	 2135427 B/op	   42262 allocs/op
```

See these trace screenshots for some visualizations of the performance data. Note that the time scale differs in each trace, and that work is properly spread across all available procs.

Screen Shot 2023-04-14 at 3 40 00 PM
Screen Shot 2023-04-14 at 3 43 26 PM

### CODE REVIEW

File: proto/merge.go [hunk 1]
```diff
@@ -73,103 +221,142 @@ func mergedFileDescriptors(debug bool) (*descriptorpb.FileDescriptorSet, error)
...
+// into a new *descriptorpb.FileDescriptorSet.
+//
+// If validate is true, do extra work to validate that import paths are properly formed
+// and that "duplicated" file descriptors across globalFiles and appFiles
+// are indeed identical, returning an error if either of those conditions are invalidated.
+func mergedFileDescriptors(globalFiles *protoregistry.Files, appFiles map[string][]byte, validate bool) (*descriptorpb.FileDescriptorSet, error) {
+	// GOMAXPROCS is the number of CPU cores available, by default.
+	// Respect that setting as the number of CPU-bound goroutines,
+	// and for channel sizes.
+	nProcs := runtime.GOMAXPROCS(0)
```
Comment (mmsqe) [inline proto/merge.go:302]: limit with CPU-bound is a good idea, but might need rename to [mergeFileDescriptors](https://github.com/cosmos/gogoproto/blob/9d031c6a9d21756d032dbdf8221474ce3674c490/proto/merge.go#L190)

Comment (mark-rushakoff) [reply proto/merge.go:302]: Good catch, thank you. Comments corrected in cf9db7151b1b6aa0b1963fcb574f7514e20fbc7e.

**Review by mmsqe — APPROVED** [review]
LGTM, thanks for the refactor.

**Review by julienrbrt — APPROVED** [review]
lgtm!

**Pull request merged**